qualtran/bloqs/basic_gates/su2_rotation.py

```diff
@@ -0,0 +1,158 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from functools import cached_property
+from typing import Any, Dict, TYPE_CHECKING
+
+import numpy as np
+from attrs import frozen
+from numpy.typing import NDArray
+
+from qualtran import bloq_example, BloqDocSpec, GateWithRegisters, Signature
+from qualtran.bloqs.basic_gates import Ry, ZPowGate
+from qualtran.cirq_interop.t_complexity_protocol import TComplexity
+from qualtran.drawing import TextBox
+
+if TYPE_CHECKING:
+    import quimb.tensor as qtn
+
+    from qualtran import BloqBuilder, Soquet, SoquetT
+    from qualtran.drawing import WireSymbol
+
+
+@frozen
+class SU2RotationGate(GateWithRegisters):
+    r"""Implements an arbitrary SU(2) rotation.
+
+    The rotation is represented by the matrix:
+
+    $$
+        e^{i \alpha}
+        \begin{pmatrix}
+        e^{i(\lambda + \phi)} \cos(\theta) & e^{i\phi} \sin(\theta) \\
+        e^{i\lambda} \sin(\theta) & - \cos(\theta)
+        \end{pmatrix}
+    $$
+
+    where $s$ is the global phase shift.
+
+    Args:
+        theta: rotation angle $\theta$ in the above matrix.
+        phi: phase angle $\theta$ in the above matrix.
+        lambd: phase angle $\lambda$ in the above matrix.
+        global_shift: phase angle $\alpha$, i.e. apply a global phase shift of $e^{i \alpha}$.
+
+    References:
+        [Generalized Quantum Signal Processing](https://arxiv.org/abs/2308.01501)
+        Motlagh and Wiebe. (2023). Equation 7.
+    """
+
+    theta: float
+    phi: float
+    lambd: float  # cannot use `lambda` as it is a python keyword
+    global_shift: float = 0
```

Comment:
Let's accept a parameter `eps` for precision and pass `eps / 3` to the 3 rotation subbloqs in the decomposition.

Reply:
The `GlobalPhase` op would also need an `eps` right? Although this matters only when controlled. I currently used `eps/4` for the four ops.